### PR TITLE
Fix imports of built-in macros (mostly)

### LINKED ISSUE #61687: Imports/reexports of built-in macros are not supported
```rust
use column; // error: cannot import a built-in macro

fn main() {}
```

The reason behind this error is that if you import something with a `DefId`, that `DefId` should be supported by all stages of compilation (metadata emitting specifically), but built-in macros don't have such a `DefId`.

The crate ID in their `DefId` has special value `CrateNum::BuiltinMacros` that is supported only inside rustc_resolve and causes ICEs every time it leaves it.

### PR DESCRIPTION
**Pull request by petrochenkov**
Built-in macros now have `DefId`s from `LOCAL_CRATE` and their `DefIndex`es occupy a continuous fixed-size range immediately following `GlobalMetaData` entries.

This way they can be used in imports without causing ICEs due to `CrateNum::BuiltinMacros`.
I tried a few different approaches mentioned in https://github.com/rust-lang/rust/issues/61687, but this one seems to be the least ugly and painful.
**EDIT**: One more alternative is to follow the example of `Res::PrimTy(hir::PrimTy)` and introduce `Res::BuiltinMacro(usize)` so that built-in macros don't need a `DefId` at all. I didn't try it yet, but perhaps I should.

Caveats:
- Same built-in macro imported through different crates has different `DefId`s.
This can be observable in corner cases (e.g. glob vs glob conflicts, see the test `import-builtin-macros-bug.rs `).
I tried to convert `(THAT_CRATE, DefIndex)` into `(LOCAL_CRATE, DefIndex)` for built-in macros immediately during decoding, but apparently there are assumptions that entries from other crates cannot have `DefId` from the current crate after decoding, so it caused ICEs.
- Individual imports of built-in macros are not stability checked (e.g. `use concat_idents;` does not produce a feature error) so the whole feature is gated instead.
The primary motivation for the fix is reexporting built-in macros from the standard library and we can use unstable features in the standard library.
The proper solution to this (and multiple other) problems is moving stability/deprecation checking into resolve.

Closes https://github.com/rust-lang/rust/issues/61687
cc https://github.com/rust-lang/rust/issues/61875
r? @eddyb

### CODE REVIEW

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -570,6 +570,9 @@ declare_features! (
     // #[repr(transparent)] on unions.
     (active, transparent_unions, "1.37.0", Some(60405), None),
 
+    // `use builtin_macro`.
```
Comment (Centril) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    // Allows `use builtin_macro;`.
```

File: src/test/ui/rust-2018/uniform-paths/import-builtin-macros-stability.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// FIXME: Individual imports of built-in macros are not stability checked right now,
+// so the whole feature is gated instead.
```
Comment (Centril) [inline src/test/ui/rust-2018/uniform-paths/import-builtin-macros-stability.rs]: I don't understand what this comment means. Can you illustrate the difference?

Comment (petrochenkov) [reply src/test/ui/rust-2018/uniform-paths/import-builtin-macros-stability.rs]: `use concat_idents as unstable;` below doesn't produce a "`concat_idents` is unstable" error.

**Comment by petrochenkov** [comment]
I'm going to try the `Res::BuiltinMacro` approach, marking as waiting-on-author.

**Comment by eddyb** [comment]
Something I really wanted years ago was to move the "definitions" of builtin types into libcore.
That is, to have `#[intrinsic = "u8"] type u8;` or `#[intrinsic = "u8"] enum u8 {}` or w/e somewhere in libcore, and have that be the canonical definition, to get rid of some of the hardcoding.
(I guess nowadays you could also abuse `extern "intrinsic" { type u8; }`)

So now I'm wondering, can we put `#[intrinsic = "concat_idents"] macro concat_idents;` in libcore?
Not a reexport of some magical item, but a definition with implementation-specified semantics.
Even if we haven't done it for builtin types yet, maybe we can do it for macros to start with?

**Comment by petrochenkov** [comment]
>So I'm wondering, can we put #[intrinsic = "concat_idents"] macro concat_idents; in libcore?

That's an interesting idea.
Attaching all this data associated with built-in macros (like stability) to an actual node in source code would make many things simpler.
I'll try it.

**Comment by petrochenkov** [comment]
For built-in types it was tried in https://github.com/rust-lang/rust/pull/32274.
(The downside is some potential inconveniences for `libcore` itself and other code using `no_core` or `no_imlicit_prelude`, but that depends on how exactly this is implemented.)

**Comment by petrochenkov** [comment]
Built-in macros introduced through `libcore` are blocked on https://github.com/rust-lang/rust/pull/61898.

File: src/test/ui/rust-2018/uniform-paths/import-builtin-macros-stability.rs (hunk 2 above)
Comment (Centril) [reply src/test/ui/rust-2018/uniform-paths/import-builtin-macros-stability.rs]: Oh I see; I read `.stderr` sloppily... thanks!

Comment (petrochenkov) [reply src/test/ui/rust-2018/uniform-paths/import-builtin-macros-stability.rs]: Ahaha, `concat_idents` and friends are actually stable *themselves*, they just report feature errors as a part of their expansion.
I'll fix this, but https://github.com/rust-lang/rust/pull/61898 is a pre-requisite.

**Comment by nikomatsakis** [comment]
@eddyb 

Nit: I feel like these should be *lang items* -- I'm not sure why we have so many categories of "special thing known to the compiler" but I think we should centralize on one of them.

**Comment by eddyb** [comment]
@nikomatsakis IMO we should distinguish between "this is exposing a compiler primitive to the library" and "this is exposing a library entity to the compiler", only the latter of which needs to be unique.

**Comment by matklad** [comment]
>Something I really wanted years ago was to move the "definitions" of builtin types into libcore.

Yess, I have been thinking exactly about this when adding primitives names resolution to rust-analyzer. It also helps with goto definition UX in IDEs, because you have the actual source to go to.

It probably shouldn't be literally libcore though, because `no_core` is a thing. Rather we need a separate builtins crate which you can't opt out of.

**Comment by mark-i-m** [comment]
> It probably shouldn't be literally libcore though, because no_core is a thing. Rather we need a separate builtins crate which you can't opt out of.

`libreallycore`

**Comment by eddyb** [comment]
> It probably shouldn't be literally libcore though, because `no_core` is a thing. Rather we need a separate builtins crate which you can't opt out of.

Nope, `#![no_core]` exists literally for `libcore` alone. It's really hard to have any other setup, because a bunch of lang items would have to be in it for anything to work, not just the builtin types/macros.

**Comment by petrochenkov** [comment]
Now waiting on https://github.com/rust-lang/rust/pull/62042

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/62086 confirmed that the idea with defining built-in macros through libcore is viable, so I'll close this PR.

**Pull request closed without merging**